source/common/network/address_impl.h

```diff
@@ -235,7 +263,26 @@ class PipeInstance : public InstanceBase {
     return sizeof(pipe_.address_);
   }
 
+  static absl::Status validateProtocolsupported() { return absl::OkStatus(); }
+
+  template <typename InstanceType, typename... Args>
+  friend InstanceType createInstance(Args&&... args);
```

Comment:
static method of whom? We have 3 types of instances here. If we make this method static, we need to implement them for all the 3 types.

Reply:
Good qusetion. You could make this a static method of InstanceBase and then you can friend InstanceBase?

But what you have is not too bad also. It does not look to me like you needed to declare the `friend`s in the other subtypes. Why is this one different?

Reply:
Need to declare friend for all types... This is still a work in progress. I haven't declared the ctors of ipv4 and ipv6 instance private.

Reply:
The problem is that a parent static method can not access the private member of children.

Reply:
It can if you friend the parent class. Another variant is to declare an `InstanceFactory` class with one static method, and friend that class where needed. That seems cleaner than friending the template function, but it's up to you.

Reply:
Thanks, a factory class would be great!